presto-native-execution/presto_cpp/main/TaskManager.cpp

```diff
@@ -1098,16 +1098,38 @@ bool TaskManager::getStuckOpCalls(
     std::vector<velox::exec::Task::OpCallInfo>& stuckOpCalls) const {
   const auto thresholdDurationMs =
       SystemConfig::instance()->driverStuckOperatorThresholdMs();
+  const auto thresholdCancelMs =
+      SystemConfig::instance()
+          ->driverCancelTasksWithStuckOperatorsThresholdMs();
+
   const std::chrono::milliseconds lockTimeoutMs(thresholdDurationMs);
   auto taskMap = taskMap_.rlock(lockTimeoutMs);
   if (!taskMap) {
     return false;
   }
-  for (const auto& [_, prestoTask] : *taskMap) {
-    if (prestoTask->task != nullptr &&
-        !prestoTask->task->getLongRunningOpCalls(
-            lockTimeoutMs, thresholdDurationMs, stuckOpCalls)) {
-      deadlockTasks.push_back(prestoTask->task->taskId());
+
+  for (const auto& [id, prestoTask] : *taskMap) {
+    if (prestoTask->task != nullptr) {
+      const auto numPrevStuckOps = stuckOpCalls.size();
+      if (!prestoTask->task->getLongRunningOpCalls(
+              lockTimeoutMs, thresholdDurationMs, stuckOpCalls)) {
+        deadlockTasks.push_back(prestoTask->task->taskId());
+      }
+      // See if we need to cancel the Task - it should be running, the cancel
+      // threshold should be valid and it should have at least one stuck driver
+      // that was stuck for enough time.
+      if (numPrevStuckOps < stuckOpCalls.size() && thresholdCancelMs != 0 &&
+          prestoTask->task->isRunning()) {
+        for (auto it = stuckOpCalls.begin() + numPrevStuckOps;
```

Comment:
why not just check all the stuck op calls? Thanks!

Reply:
This is because we collect stuck opcalls from many tasks in stuckOpCalls and only the last (size - numPrevStuckOps) belong to our current task.